baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ImmutablesBuilderMissingInitialization.java

```diff
@@ -0,0 +1,315 @@
+ * check will stop working (but the check will probably pass).
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "ImmutablesBuilderMissingInitialization",
+        linkType = LinkType.CUSTOM,
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "All required fields of an Immutables builder must be initialized")
+public final class ImmutablesBuilderMissingInitialization extends BugChecker implements MethodInvocationTreeMatcher {
+
+    private static final String FIELD_INIT_BITS_PREFIX = "INIT_BIT_";
+
+    /**
+     * Prefixes on the interface getter methods that may be stripped by immutables.
+     *
+     * The default immutables style just uses get*, but some places use is* too and we can't load the style to check
+     * what to remove. It doesn't matter if we remove too much, because we only do a suffix match on the methods.
+     *
+     * This should only help when the unprefixed field name is an identifier, because in other cases immutables already
+     * uses the unprefixed name.
+     */
+    private static final ImmutableSet<String> GET_PREFIXES = ImmutableSet.of("GET_", "IS_");
+
+    private static final Matcher<ExpressionTree> builderMethodMatcher = Matchers.instanceMethod()
+            .onClass(ImmutablesBuilderMissingInitialization::extendsImmutablesGeneratedClass)
+            .named("build")
+            .withParameters();
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        // Check that the method is a builder's build() method
+        if (!builderMethodMatcher.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+        ClassSymbol builderClass = ASTHelpers.enclosingClass(ASTHelpers.getSymbol(tree));
+        ClassSymbol immutableClass = ASTHelpers.enclosingClass(builderClass);
+        if (immutableClass == null) {
+            return Description.NO_MATCH;
+        }
+        Optional<ClassSymbol> interfaceClass = Streams.concat(
+                        immutableClass.getInterfaces().stream(), Stream.of(immutableClass.getSuperclass()))
+                .map(type -> type.tsym)
+                .map(filterByType(ClassSymbol.class))
+                .flatMap(Streams::stream)
+                .filter(classSymbol ->
+                        ASTHelpers.hasAnnotation(classSymbol, "org.immutables.value.Value.Immutable", state))
+                .findAny();
+        if (!interfaceClass.isPresent()) {
+            return Description.NO_MATCH;
+        }
+
+        // Mandatory fields have a private static final constant in the generated builder named INIT_BIT_varname, where
+        // varname is the UPPER_UNDERSCORE version of the variable name. Find these fields to get the mandatory fields.
+        // nb. this isn't part of any immutables API, so could break.
```

Comment:
this is kinda amazing that it works! given that it relies on internal implementation details of immutables, I wonder if we get this error-prone rule upstreamed into https://github.com/immutables/immutables so that if they ever decide to change the internal representation then CI will catch their change??

Reply:
One concern I would have with upstreaming this is that it's not totally compliant with custom `@Value.Style` annotations, and can never be because the annotation (eg `@ImmutablesStyle` internally) is often brought in as a `compileOnly` dependency, so if it's consumed from a different compilation unit (eg using immutable classes that were generated in the main sourceset from test code) it will not be possible to recover the original style. That only affects an edge case where the `get` style has been customized and the field with that part removed is an identifier (eg `getPublic`), so maybe it's not a big deal (I solved it here by hard coding the style to `get*` and `is*`and bailing if there are fields that can't be initialized after transforming like that).

Instead, if it were upstreamed then I'd prefer to go with an annotations-based approach, as mentioned above.